src/main/java/com/mojang/datafixers/types/Type.java

```diff
@@ -157,11 +159,33 @@ public TypeTemplate template() {
     @SuppressWarnings("unchecked")
     public Optional<RewriteResult<A, ?>> rewrite(final TypeRewriteRule rule, final PointFreeRule fRule) {
         final Triple<Type<?>, TypeRewriteRule, PointFreeRule> key = Triple.of(this, rule, fRule);
-        if (!REWRITE_CACHE.containsKey(key)) {
-            final Optional<? extends RewriteResult<?, ?>> result = rule.rewrite(this).flatMap(r -> r.view().rewrite(fRule).map(view -> RewriteResult.create(view, r.recData())));
+        // This code under contention would generate multiple rewrites, so we use CompletableFuture for pending rewrites.
+        // We can not use computeIfAbsent because this is a recursive call that will block server startup
+        // during the Bootstrap phrase that's trying to pre cache these rewrites.
+        Optional<? extends RewriteResult<?, ?>> rewrite = REWRITE_CACHE.get(key);
+        //noinspection OptionalAssignedToNull
+        if (rewrite != null) {
+            return (Optional<RewriteResult<A, ?>>) rewrite;
```

Comment:
Cast should not be neccesary if the map stores `CompletableFuture<? extends Optional<? extends RewriteResult<?, ?>>>>`

Reply:
just noticed the cache key references this anyways... I'm just going to move it to an instance property

Reply:
I think the reason you failed is that you need another static field with an object that you sync around instead of the instance cache, isn't it?